go/src/lib.rs

```diff
@@ -170,11 +187,166 @@ pub unsafe extern "C" fn close_client(client_adapter_ptr: *const c_void) {
     connection_response_ptr: *mut ConnectionResponse,
 ) {
-    assert!(connection_response_ptr.is_null());
+    assert!(!connection_response_ptr.is_null());
     let connection_response = unsafe { Box::from_raw(connection_response_ptr) };
     let connection_error_message = connection_response.connection_error_message;
     drop(connection_response);
     if !connection_error_message.is_null() {
         drop(unsafe { CString::from_raw(connection_error_message as *mut c_char) });
     }
 }
+
+/// Deallocates a `CommandResponse`.
+///
+/// This function also frees the contained string_value and array_value. If the string_value and array_value are null pointers, the function returns and only the `CommandResponse` is freed.
+///
+/// # Panics
+///
+/// This function panics when called with a null `CommandResponse` pointer.
+///
+/// # Safety
+///
+/// * `free_command_response` can only be called once per `CommandResponse`. Calling it twice is undefined behavior, since the address will be freed twice.
+/// * `command_response_ptr` must be obtained from the `CommandResponse` returned in [`success_callback`] from [`command`].
+/// * `command_response_ptr` must be valid until `free_command_response` is called.
+/// * The contained `string_value` must be obtained from the `CommandResponse` returned in [`success_callback`] from [`command`].
+/// * The contained `string_value` must be valid until `free_command_response` is called and it must outlive the `CommandResponse` that contains it.
+#[no_mangle]
+pub unsafe extern "C" fn free_command_response(command_response_ptr: *mut CommandResponse) {
+    assert!(!command_response_ptr.is_null());
+    let command_response = unsafe { Box::from_raw(command_response_ptr) };
+    let string_value = command_response.string_value;
+    let string_value_len = command_response.string_value_len;
+    drop(command_response);
+    if !string_value.is_null() {
+        let len = string_value_len as usize;
+        Vec::from_raw_parts(string_value, len, len);
+    }
+}
+
+// TODO: Finish documentation
+/// Converts a double pointer to a vec.
+///
+/// # Safety
+///
+/// * TODO: finish safety section.
+pub unsafe fn convert_double_pointer_to_vec(
+    data: *const *const c_char,
+    len: usize,
+) -> Result<Vec<String>, Utf8Error> {
+    from_raw_parts(data, len)
+        .iter()
+        .map(|arg| CStr::from_ptr(*arg).to_str().map(ToString::to_string))
+        .collect()
+}
+
+pub fn convert_vec_to_pointer<T>(mut vec: Vec<T>) -> (*mut T, c_long) {
+    vec.shrink_to_fit();
+    let vec_ptr = vec.as_mut_ptr();
+    let len = vec.len() as c_long;
+    mem::forget(vec);
+    (vec_ptr, len)
+}
+
+// TODO: Finish documentation
+/// Executes a command.
+///
+/// # Safety
+///
+/// * TODO: finish safety section.
+#[no_mangle]
+pub unsafe extern "C" fn command(
+    client_adapter_ptr: *const c_void,
+    channel: usize,
+    command_type: RequestType,
+    arg_count: usize,
+    args: *const *const c_char,
```

Comment:
Similar to "CommandsReponse", args[0] might hold a null value as a valid value: (think of "set file <PNG file content>")

You should probably use the same approach you used in `CommandResponse` and pass 3 arguments here:

```rust
pub unsafe extern "C" fn command(
    client_adapter_ptr: *const c_void,
    channel: usize,
    command_type: RequestType,
    arg_count: usize,
    args: *const *const c_char,
    args_len: *const c_long,
```

Reply:
Done